lib/propolis/src/hw/testdev.rs

```diff
@@ -0,0 +1,113 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Devices intended for testing purposes.
+//!
+//! These devices do things which are generally unwanted in real life, such as
+//! "intentionally breaking Propolis", "intentionally breaking the guest OS", or
+//! some combination of the two.
+
+use crate::inventory::Entity;
+use crate::migrate::*;
+
+use serde::{Deserialize, Serialize};
+
+use slog::info;
+use std::sync::{
+    atomic::{AtomicUsize, Ordering},
+    Arc,
+};
+
+/// A test device for simulating migration failures.
+pub struct MigrationFailureDevice {
+    log: slog::Logger,
+    exports: AtomicUsize,
+    fail: MigrationFailures,
+}
+
+pub struct MigrationFailures {
+    pub exports: usize,
+    pub imports: usize,
+}
+
+#[derive(Clone, Default, Deserialize, Serialize)]
+struct MigrationFailurePayloadV1 {
+    /// If set, an attempt to import this device should fail.
+    fail_import: bool,
+}
```

Comment:
Putting this in the payload (if I'm understanding things correctly) means that the configuration passed to the migration *source* determines whether a failure gets injected into the *destination*. The code's correct, but this feels a bit unintuitive to me--if/when we add more failure injection capabilities I reckon most of them will need to be configured directly on the process we want to fail. Is there another advantage to having the source dictate this behavior that I'm overlooking?

Reply:
Hm, yeah, that's a good point. The reason that it works this way is mostly historical: when I wrote this code, I hadn't yet decided that the failure injection device would be configured only though the `config.toml` and not through `InstanceEnsure`, so it hadn't occurred to me that there would be a _way_ to configure the device in the migration target except by configuring it in the source and having that tell the target what to do. I can change this so that the import failure configuration is set in the target!

Reply:
Addressed this in 3705717b3ae3b8944631f6809783530cd49f3e9e !